src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsStream.cs

```diff
@@ -0,0 +1,181 @@
+
+        public TdsStream(Stream underLyingStream) : base()
+        {
+        }
+
+        /// <summary>
+        /// Replaces the underlying stream. This is useful while changing between SSL stream and non-SSL stream. 
+        /// e.g. Prelogin ends with TLS handshake, which is done on a different stream. Once the handshake is done,
+        /// login is sent in the TLS stream. However at the end of the login exchange, the stream may be switched 
+        /// back to the non-SSL stream, which is either the pipe stream or the TCP stream.
+        /// </summary>
+        /// <param name="stream"></param>
+        /// <exception cref="NotImplementedException"></exception>
+        public void ReplaceUnderlyingStream(Stream stream)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// When writing the packet, the caller needs to 
+        /// specify the packet type. 
+        /// TODO: Consider accepting an enum of packet types
+        /// instead of the byte.
+        /// </summary>
+        /// <param name="packetType"></param>
+        public void SetWritePacketType(TdsStreamPacketType packetType)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void Flush()
+        {
+            throw new NotImplementedException();
+        }
+
+        public override int Read(Span<byte> buffer)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override int Read(
+            byte[] buffer, 
+            int offset, 
+            int count)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override ValueTask WriteAsync(
+            ReadOnlyMemory<byte> buffer,
+            CancellationToken cancellationToken)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// We only support ValueTask based overloads for async IO.
+        /// This overload makes sure that the method throws, instead 
+        /// of using the unoptimized path.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="offset"></param>
+        /// <param name="count"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotSupportedException"></exception>
+        public override Task WriteAsync(
+            byte[] buffer,
```

Comment:
This will never be implemented. The ReadOnlyMemory overload with ValueTask is what this stream would expose, which would call into underlying's stream's ReadOnlyMemory -> ValueTask overload.

Reply:
I have explicitly called out a NotSupportedException() in this overload, so that the default overrides of this method in the base class dont allow for unoptimized behavior.

Reply:
> allow for unoptimized behavior.

I don't think the byte[] overload  is "unoptimized" behavior, just not necessarily the most optimal behavior. For instance, if I have a constantly defined byte array and I need to write it, it is optimal to send the byte array directly. However, if we are forcing the ReadOnlyMemory overload, that means we will always need to new up a Memory struct before writing. And worse, if we are on net48, we will need to revert it back to a byte[] when we call any underlying stream. I think offering both options is better than forcing one or the other. Especially since we are effectively saying to the consumer "here is a Stream ... it might not actually implement all the things a Stream does, tho."

The approach I took for the MarsSessionStream and MarsSessionMuxDemux was to conditionally add the ReadOnlyMemory overload if we're on net8 or greater. I think this is best because it allows passing calls directly through the stream chain. The downside to it is that any consumer needs to also switch behavior based on dotnet version (at least where either call could be used, which I'm not sure will be very many situations).

Reply:
@benrr101  and @cheenamalhotra  you are right. 
I agree with the comment of byte[] overload not being unoptimized. I am hoping to stay away from Task based overloads.
Span and Memory are castable to byte[] so if the caller has a byte[], then it can call the stream to write the byte.

For Netfx @benrr101  and I discussed, since ValueTask is available in an extension Nuget package, for .Net framework, the stream class will wrap the underlying stream's Task into the ValueTask and expose them. This would make sure that we stay as performant as possible where ValueTask is available, but at the same time, with the same APIs we can server NetFx as well. 

I will be happy to revisit this point, if we see clients which absolutely need Task based overloads. That wont be too hard to expose.

Reply:
Sounds good, but yes, ideally we should ensure when the code comes in, we support all APIs possible to provide complete support for all targets.